include/runtime/collections/RBTree.h

```diff
@@ -0,0 +1,606 @@
+              lft.right().left()),
+          lft.right().root_key(), lft.right().root_val(),
+          RBTree(Color::B, lft.right().right(), x, v, rgt));
+    } else if (rgt.doubled_left()) {
+      return RBTree(
+          minus_one_color(c), RBTree(Color::B, lft, x, v, rgt.left().left()),
+          rgt.left().root_key(), rgt.left().root_val(),
+          RBTree(
+              Color::B, rgt.left().right(), rgt.root_key(), rgt.root_val(),
+              rgt.right()));
+    } else if (rgt.doubled_right()) {
+      return RBTree(
+          minus_one_color(c), RBTree(Color::B, lft, x, v, rgt.left()),
+          rgt.root_key(), rgt.root_val(), rgt.right().paint(Color::B));
+    } else {
+      return RBTree(c, lft, x, v, rgt);
+    }
+  }
+
+  bool is_empty(Color c) const { return is_empty() && leaf_color() == c; }
+
+  bool non_empty(Color c) const { return !is_empty() && root_color() == c; }
+
+  bool singleton(Color c) const {
+    return !is_empty() && root_color() == c && left().is_empty(Color::B)
+           && right().is_empty(Color::B);
+  }
+
+  bool only_left_child(Color parent_color, Color child_color) const {
+    return !is_empty() && root_color() == parent_color && !left().is_empty()
+           && left().root_color() == child_color && right().is_empty(Color::B);
+  }
+
+  bool only_right_child(Color parent_color, Color child_color) const {
+    return !is_empty() && root_color() == parent_color
+           && left().is_empty(Color::B) && !right().is_empty()
+           && right().root_color() == child_color;
+  }
+
+  bool doubled_left() const {
+    return !is_empty() && root_color() == Color::R && !left().is_empty()
+           && left().root_color() == Color::R;
+  }
+
+  bool doubled_right() const {
+    return !is_empty() && root_color() == Color::R && !right().is_empty()
+           && right().root_color() == Color::R;
+  }
+
+  RBTree paint(Color c) const {
+    assert(!is_empty());
+    return RBTree(c, left(), root_key(), root_val(), right());
+  }
+
+private:
+  std::shared_ptr<const Node> root_;
+};
+
+// Recursively (using inorder traversal) apply function f to all elements of
+// tree t. Function f must accept two arguments of types T and V respectively.
+template <class T, class V, class F>
+void for_each(RBTree<T, V> const &t, F &&f) {
+  if (!t.is_empty()) {
+    for_each(t.left(), std::forward<F>(f));
+    std::invoke(f, t.root_key(), t.root_val());
+    for_each(t.right(), std::forward<F>(f));
+  }
+}
+
+// Return a red-black tree with all elements in t, and then also from the
+// container designated by the beginning and end iterator arguments. The
+// container should contain elements of type std::pair<T,V>.
+template <class T, class V, class I>
+RBTree<T, V> inserted(RBTree<T, V> const &t, I it, I end) {
+  if (it == end) {
+    return t;
+  }
+  T key = it->first;
+  V val = it->second;
+  auto t1 = t.inserted(key, val);
+  return inserted(t1, ++it, end);
+}
```

Comment:
I'm not keen on these getting leaked into the global namespace every time someone includes this header; the structure code could be kept in an `rb_tree` namespace or similar.

Reply:
The structure is now enclosed in namespace `rb_tree`.